### ISSUE TITLE
Component helper not updating when component name changes

### ISSUE THREAD

**Issue by nlfurniss** [emberjs/ember.js#16832]
I’m upgrading my app from Ember 2.18.2 to 3.1.2 and I’m running into an issue with the `component` helper and yielding components.

Here is a [Twiddle demonstrating the issue](https://ember-twiddle.com/d65fa55ddb07dade0afffcb4256ee48b). When the Ember dependency is set to 2.18.2 the text changes. When you set Ember to 3.0.0 the text changes. When Ember is set to 3.1.3 the text does not update.

For anyone else running into this issue, Tom Dale provided a workaround in [this Twiddle](https://ember-twiddle.com/0fb1fe1f6428d554a077efe08ee017c4?openFiles=templates.index.hbs%2C).

Rob Jackson:
```
Tom and I dug into it a bit and came up with the same work around (using the component helper). The issue comes down to this AST transform completely failing when there are no positional or named arguments to an invocation:

https://github.com/emberjs/ember.js/blob/v3.4.0-beta.1/packages/ember-template-compiler/lib/plugins/transform-dot-component-invocation.ts#L88
```

**Comment by pixelhandler** [comment]
@nlfurniss I see you have a workaround now, thanks for sharing the solution and pointing out the issue with "no positional or named arguments".

Feel free to close this when ready

**Comment by nlfurniss** [comment]
@pixelhandler why did you close this? There is a workaround, but it's a regression in Ember that should be fixed

**Comment by rwjblue** [comment]
Whoops!

**Comment by pixelhandler** [comment]
@nlfurniss just checking back, is this resolved in a current release of Ember ?

**Comment by nlfurniss** [comment]
No it has not

**Comment by chancancode** [comment]
(I don't think the bug/fix is really about the AST transform, those are just "opportunistic" – when there are syntactic positions that are known to be component invocations we hard coded them to be such, but this should work)

**Comment by chancancode** [comment]
This is a simplified test case: https://ember-twiddle.com/5f7ed4b003e14ef40e47f8de1aea5be6

**Comment by chancancode** [comment]
If anyone has time, would be good to turn that simplified test case into a failing test, both here and in glimmer-vm.

**Comment by lifeart** [comment]
@rwjblue looks like issue covered in https://github.com/emberjs/ember.js/blob/master/packages/%40ember/-internals/glimmer/tests/integration/components/dynamic-components-test.js#L32

**Comment by rwjblue** [comment]
@lifeart - That is a reasonable _example_ test case, but it does not test what this issue is reporting. Another test would need to be added that has an invocation like `{{#let (component this.someThing) as |comp|}}{{comp}}{{/let}}` (no other positional or named arguments).

**Comment by lifeart** [comment]
@rwjblue lol, looks like test failing https://github.com/emberjs/ember.js/pull/19146

**Comment by chriskrycho** [comment]
Just hit this again today. Will see if I can *maybe* find my way to a solution by hacking on/pushing at this tomorrow.

**Comment by chriskrycho** [comment]
Update: I spent some time poking at it but lack the relevant expertise. For the moment, we're just working around it more.

**Comment by johanrd** [comment]
**TLDR: this seems to no longer be a problem if you use angle bracket invocation `<MyComponent>` instead of `{{curly}}`:**

I tried to reproduce now, and I am happy to announce that at least one of the two tests passes now: 

- `{{comp}}` (curly) — **broken**, doesn't re-render when name changes (see [test failure](https://github.com/johanrd/ember.js/actions/runs/22807979720/job/66160117193?pr=3#step:5:18))
- `<Comp` (angle bracket) — **works fine** (see passing test suite: https://github.com/johanrd/ember.js/pull/4)

So the `(component)` helper itself seems to track reactivity correctly. It's only curly invocation of the yielded result that fails to pick up the update.

…Since curly component invocation is the legacy/discouraged syntax and angle brackets are the modern way, this is arguably low-priority. The fix for any user hitting this is just: use `<Comp />` instead of `{{comp}}`. Might be worth closing https://github.com/emberjs/ember.js/issues/16832 and https://github.com/emberjs/ember.js/pull/19146 with that note rather than fixing deep Glimmer VM curly invocation reactivity.?
